File: MissingElement.cs
namespace Algorithms
{
    // Did this code successfully run on Leetcode : Could not find the specific problem on leetcode, but ran this on my system to verify
    // Any problem you faced while coding this : It took a while to come up with binary search solution. I started with wrong condition
    // for checking the mid and ran into issues. 
    public class MissingElement
    {
        //Brute Force Approach
        //Time Complexity - O(n) - As we are iterating throught the whole array once atleast
        //Space Complexity - O(1) - Since we are not using any extra space
        public int FindMissingElement(int[] nums)
        {
            for (int i = 0; i < nums.Length; i++)
            {
                if (!(nums[i] == nums[i + 1] - 1))
                {
                    return nums[i + 1] - 1;
                }
            }
            return -1;
        }

        //Using extra array
        //Time Complexity - 2O(n) which eventually goes down to O(n) - Once for iterating through new array and old array
        //Space Complexity - O(n) - Since we have taken additional array for comparision
        public int FindMissingElementExtraArray(int[] nums)
        {
            var lastElem = nums[nums.Length - 1];

            var tempArr = new int[lastElem];

            //Initialize the temp array
            for (int i = 0; i < nums.Length; i++)
            {
                tempArr[i] = i + 1;
            }

            //Compare the temp array and return the first mismatched element
            for(int i = 0; i < nums.Length; i++)
            {
                if (nums[i] != tempArr[i]) return tempArr[i];
            }
            return -1;
        }

        //Using Binary Search
        //Time Complexity - O(logn) - Since we are reducing the search space by half
        //Space Complexity - O(1) - Since we are not using any extra space
        public int FindMissingElementBinarySearch(int[] nums) //[1, 2, 3, 5, 6, 7]
        {
            var left = 0;
            var right = nums.Length - 1;
            while(left <= right)
            {
                var mid = left + (right - left) / 2;
                // If mid is not equal the element in mid, then that means the missing element lies on the left part of the array
                if (nums[mid] != mid)
                {                    
                    if(nums[mid-1] + 1 != mid)
                    {
                        return mid;
                    }
                    //Since we are searching on the left part of the array
                    right = mid - 1;
                }
                else
                {
                    left = mid + 1;
                }
            }
            return -1;
        }
    }
}

File: missingElement.java
//Missing Element problem- Competitive coding problem 
// Time Complexity : O(log n) because we divided the array into half and kept searching in it. In short we straight away reducing the half part of array
// Space Complexity : O(1) as we are not storing 
// Any problem you faced while coding this : no

/*Output
The missing number is:6

*/

public class missingElement {

	private static int missingElements(int[] arr, int arr_length) {
		//apply Binary Search on arr to find the missing element. 
		
		int low=0;
		int high=arr_length-1;
		
		//check difference between arr[mid] and it'd index i.e. mid and if it's equal to arr[low] and low then move towards right i.e. low = mid+1
		//if not then vise versa

		while(low<high) {
		    int mid=low+((high-low)/2);
//if missing element is at mid
			if((arr[mid] - mid) != (arr[mid-1]-(mid-1)))
			{
				int r=(arr[mid-1]+arr[mid])/2;
				return r;
			}
			else if((arr[mid] - mid) != (arr[mid+1]-(mid+1)))
			{
				int r=(arr[mid+1]+arr[mid])/2;
				return r;
			}

			if((arr[mid]-mid) == (arr[low]-low))
			{
				low=mid+1;
			}
			else //if((arr[mid]-mid) != (arr[low]-low))
			{
				high=mid-1;
			}
		}
		return 0;
	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub
	//	int[] arr= {1,2,3,5,6,7};
		int[] arr= {1,2,3,4,5,7};
	//	int[] arr= {1,3,4,5,6,7};
	//	int[] arr= {2,3,4,5,6,7};
		
		int result=missingElements(arr, arr.length);
		System.out.println("The missing number is:" + result);
	}

	
}
